server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts

```diff
@@ -1,39 +1,110 @@
+} from '@amzn/codewhisperer-streaming'
+import { ChatParams, ErrorCodes, ResponseError } from '@aws/language-server-runtimes/protocol'
 import {
     CancellationToken,
-    ChatParams,
     CredentialsProvider,
     EndChatParams,
     Server,
+    TabAddParams,
+    TabRemoveParams,
 } from '@aws/language-server-runtimes/server-interface'
+import { ChatEventParser } from './chat/chatEventParser'
 import { ChatSessionManagementService } from './chat/chatSessionManagementService'
+import { convertChatParamsToRequestInput } from './chat/utils'
 
 export const QChatServer =
     (service: (credentialsProvider: CredentialsProvider) => ChatSessionManagementService): Server =>
     ({ chat, credentialsProvider, logging }) => {
-        const codewhispererclient: ChatSessionManagementService = service(credentialsProvider)
+        const chatSessionManagementService: ChatSessionManagementService = service(credentialsProvider)
+
+        chat.onTabAdd((params: TabAddParams) => {
+            logging.log('Received tab add request')
+
+            // while createSession should throw an error when we are trying to
+            // create a duplicate session, but this handler doesn't need to care
+            if (chatSessionManagementService.getSession(params.tabId)) {
+                logging.log('Session already exists')
+            } else {
+                chatSessionManagementService.createSession(params.tabId)
+            }
+        })
+
+        chat.onTabRemove((params: TabRemoveParams) => {
+            logging.log('Received tab remove request')
+            chatSessionManagementService.deleteSession(params.tabId)
+        })
 
         chat.onEndChat((params: EndChatParams, _token: CancellationToken) => {
-            logging.log('received end chat request')
-            codewhispererclient.deleteSession(params.tabId)
+            logging.log('Received end chat request')
+
+            chatSessionManagementService.deleteSession(params.tabId)
 
             return true
         })
 
         chat.onChatPrompt(async (params: ChatParams, token: CancellationToken) => {
-            logging.log('received chat prompt')
+            logging.log('Received chat prompt')
 
-            const session = codewhispererclient.getSession(params.tabId)
+            const session = chatSessionManagementService.getSession(params.tabId)
+
+            if (!session) {
+                logging.log(`Session not found for tabId: ${params.tabId}`)
+                return new ResponseError(ErrorCodes.InvalidParams, 'Session not found')
+            }
 
             token.onCancellationRequested(() => {
+                logging.log('cancellation requested')
                 session.abortRequest()
             })
 
-            return {}
+            let requestInput: GenerateAssistantResponseCommandInput
+            let response: GenerateAssistantResponseCommandOutput
+
+            try {
+                requestInput = convertChatParamsToRequestInput(params)
```

Comment:
You could return Result<T> from `convertChatParamsToRequestInput`. Then you would not need to throw error there, catch it here and transform into error. It would look a bit better.

Reply:
Depending on how your other error handling logic for chat looks like and if you want to reuse that `Result` in other places, you could return `ResponseError` there or more generic error (like string, for example) and transform it to `ResponseError` in `onChatPrompt`.

Reply:
Thanks for the suggestion and the example provided offline. I have updated to use the result pattern